openhcl/hcl/src/ioctl/x64.rs

```diff
@@ -176,29 +173,27 @@ impl<'a> BackingPrivate<'a> for MshvX64<'a> {
         let BackingState::Mshv { reg_page } = &vp.backing else {
             return Err(NoRunner::MismatchedIsolation);
         };
-        let this = if let Some(sidecar) = sidecar {
-            // Sidecar always provides a register page, but it may not actually
-            // be mapped with the hypervisor. Use the sidecar's register page
-            // only if the mshv_vtl driver thinks there should be one.
-            Self {
-                reg_page: reg_page
-                    .is_some()
-                    .then(|| NonNull::new(sidecar.register_page()).unwrap()),
-                cpu_context: NonNull::new(sidecar.cpu_context().cast()).unwrap(),
-                source: PhantomData,
-            }
-        } else {
-            Self {
-                reg_page: reg_page.as_ref().map(|x| NonNull::new(x.as_ptr()).unwrap()),
-                cpu_context: NonNull::new(
-                    // SAFETY: The run page is guaranteed to be mapped and valid.
-                    unsafe { std::ptr::addr_of_mut!((*vp.run.as_ptr()).context) }.cast(),
-                )
-                .unwrap(),
-                source: PhantomData,
-            }
-        };
-        Ok(this)
+
+        // SAFETY: The run page and register page, whether provided locally
+        // or by sidecar, are guaranteed to be mapped and valid.
+        unsafe {
+            let this = if let Some(sidecar) = sidecar {
+                // Sidecar always provides a register page, but it may not actually
+                // be mapped with the hypervisor. Use the sidecar's register page
+                // only if the mshv_vtl driver thinks there should be one.
+                Self {
+                    reg_page: reg_page.is_some().then(|| &*sidecar.register_page().cast()),
+                    cpu_context: &*sidecar.cpu_context().cast(),
+                }
+            } else {
+                Self {
+                    reg_page: reg_page.as_ref().map(|x| x.as_ref()),
+                    cpu_context: &*(&raw const (*vp.run.as_ptr()).context).cast(),
```

Comment:
I think `&raw mut` may be more appropriate here. I am not certain if it matters here, but in some cases it does (e.g., if you use a `&T` and later cast it to a `*mut T` and then to a `&mut T`, miri complains, even if you have exclusive access to the `T`).

And since there's no interior mutability in the `T` here (we are applying it on the outside via the `UnsafeCell` cast), `&raw mut` is more descriptive of what's going on.

Same elsewhere where you use `addr_of` instead of `addr_of_mut` (but also, maybe just switch those to `&raw mut`, since they are starting to deprecate `addr_of`).